pkg/util/watch.go

```diff
@@ -409,10 +359,72 @@ func Watch[C WatchClient[L], L metav1.ListMetaAccessor, T any, P WatchObject[T]]
 		}
 	}()
 
-	return &store, nil
+	return store, nil
+}
+
+type ProcessEventArgs[T any, P WatchObject[T]] struct {
+	Name   string
+	UID    types.UID
+	Config WatchConfig
+	Event  watch.Event
+	Object P
 }
```

Comment:
I think it's probably cleaner to move the rest of the post-error-check handling into `ProcessEvent`, which allows `Name`, `UID`, and `Object` to be derived inside. With that, we probably don't need `ProcessEventArgs`?

Reply:
I don't understand what you mean by `post-error-checking`. 

The `goto` and loops make it hard to move more around without making a larger (and therefore riskier) cut. I think as an intermediate state.